### PR TITLE
libMesh update: Add an optional parameter in executioner for MatSNESMFGetReuseBase.

### LINKED ISSUE #12361: SNESMFReuseBase
## Rationale
Add an optional parameter for executioner for https://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/SNES/MatSNESMFGetReuseBase.html. Some user cases such as contact problems sometime can not reuse the base vector because the problem configuration may be 
changed during Jacobian evaluation so that F(x_n) does not equal to MF(x_n). 

## Description
Add one option in executioner in case we turn it off when the user case does not satisfy the assumption (F(x_n) = MF(x_n)).   Also turn it off for contact problems. 

## Impact
Everything should stay fine. Does not impact anybody.

### PR DESCRIPTION
**Pull request by fdkong**
Some user cases such as contact problems sometime can not reuse the base vector
because the problem configuration may be changed during Jacobian evaluation
so that F(x_n) does not equal to MF(x_n).

Closes #12361

### CODE REVIEW

**Comment by fdkong** [comment]
Waiting for libmesh update

File: framework/src/executioners/Executioner.C [hunk 1]
```diff
@@ -145,6 +149,8 @@ Executioner::Executioner(const InputParameters & parameters)
       getParam<bool>("compute_initial_residual_before_preset_bcs");
 
   _fe_problem.getNonlinearSystemBase()._l_abs_step_tol = getParam<Real>("l_abs_step_tol");
+
+  _fe_problem.setSNESMFReuseBase(getParam<bool>("snesmf_reuse_base"));
```
Comment (lindsayad) [inline framework/src/executioners/Executioner.C]: You could add an overload of `FEProblemBase::setSNESMFReuseBase` that takes two arguments, with the second argument specifying whether the parameter was set by the user. So the method you would want to use is [`InputParameters::isParamSetByUser`](http://mooseframework.org/docs/doxygen/moose/InputParameters_8C_source.html#l00793). So here if you go with the overload, you could do:
```c++
_fe_problem.setSNESMFReuseBase(getParam<bool>("snesmf_reuse_base"), parameters.isParamSetByUser("snesmf_reuse_base"));
```

Comment (fdkong) [reply framework/src/executioners/Executioner.C]: I see.

**Comment by jwpeterson** [comment]
@fdkong, please add the following to your commit message for this change:

Brief summary of changes in this update:
* Fix trailing return type issue affecting GCC 7.1
* Optimization: acquire BC lists outside the CheckPointIO id loop.
* Add --enable-metaphysicl-required configure option.
* Add configure test for std containers with iterating-returning erase.
* Avoid modifying solution vector in TaoOptimizationSolver callbacks.
* Fix warnings discovered in GCC 8.1 and clang 6.1.
* Fix bug in Parallel::minloc/maxloc.
* Add ability to disable default GhostingFunctors.
* Upgrade libMesh MetaPhysicL submodule to version 0.3.1.
* Add new PetscNonlinearSolver interface: set_snesmf_reuse_base().
* Make SparseMatrix::add() const correct.

**Comment by jwpeterson** [comment]
The "Distributed mesh sweep" test failure is only one test that changed recently and the diff is bogus... I will try to push up a fix for that.

**Comment by lindsayad** [comment]
The "App tests PETSc alt" BISON single test failure is exactly the failure we expect. Once the new `Executioner` option is available (once this PR is merged) we will fix that test.

**Review by jwpeterson — APPROVED** [review]

**Review by lindsayad — APPROVED** [review]

**Pull request merged**